File: gateway/kotlin/gateway/src/main/kotlin/dev/bridgething/gateway/FrameAccumulator.kt
package dev.bridgething.gateway

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

/**
 * Per-stream buffer that takes raw byte chunks and yields complete frames.
 *
 * Bytes from a stream-oriented transport (RFCOMM, BluetoothSocket, sockets)
 * arrive without respect to the bridgething frame boundary. [FrameAccumulator]
 * keeps a rolling buffer, validates each header as soon as 16 bytes are
 * available, waits for the full payload, and pops one complete frame at a
 * time.
 *
 * Caller-driven: feed bytes with [append], drain with repeated [nextFrame]
 * until it returns null. Not thread-safe; the gateway owns one per device and
 * only calls it from inside its own mutex.
 */
public class FrameAccumulator(
  public val maxPayloadSize: Int = DEFAULT_MAX_PAYLOAD_SIZE,
) {
  public companion object {
    /**
     * Hard ceiling for an individual frame's payload. Defends against a
     * hostile peer claiming a multi-gigabyte payload before any bytes
     * actually arrive.
     */
    public const val DEFAULT_MAX_PAYLOAD_SIZE: Int = 8 * 1024 * 1024
  }

  public sealed class Exception(message: String) : RuntimeException(message) {
    public class InvalidMagic(public val magic: Int) :
      Exception("invalid magic 0x${magic.toString(16)}")
    public class UnsupportedVersion(public val version: Byte) :
      Exception("unsupported wire version $version")
    public class UnsupportedCompression(b: Byte) :
      Exception("unknown compression byte 0x${(b.toInt() and 0xff).toString(16)}")
    public class UnsupportedEncoding(b: Byte) :
      Exception("unknown encoding byte 0x${(b.toInt() and 0xff).toString(16)}")
    public class PayloadTooLarge(public val payloadLength: Long, public val max: Int) :
      Exception("payload $payloadLength exceeds max $max")
  }

  private val buffer = ByteArrayOutputStream()

  public fun append(chunk: ByteArray) {
    buffer.write(chunk)
  }

  /**
   * Pops one complete frame from the head of the buffer if available. Returns
   * null when the buffer doesn't yet contain a full header + payload. Throws
   * on bad magic, unsupported header bytes, or oversized payloads — the
   * caller is expected to drop the connection in those cases since the stream
   * has lost framing and there's no safe resync.
   */
  public fun nextFrame(): ByteArray? {
    val bytes = buffer.toByteArray()
    if (bytes.size < FrameHeader.LENGTH) return null

    val header = ByteBuffer.wrap(bytes, 0, FrameHeader.LENGTH)
    val magic = header.short.toInt() and 0xffff
    if (magic != FrameHeader.MAGIC) throw Exception.InvalidMagic(magic)
    val version = header.get()
    if (version != FrameHeader.VERSION) throw Exception.UnsupportedVersion(version)
    val compressionByte = header.get()
    if (Compression.entries.none { it.byte == compressionByte }) {
      throw Exception.UnsupportedCompression(compressionByte)
    }
    val encodingByte = header.get()
    if (Encoding.entries.none { it.byte == encodingByte }) {
      throw Exception.UnsupportedEncoding(encodingByte)
    }
    header.position(header.position() + 3) // reserved
    val payloadLen = header.long
    if (payloadLen > maxPayloadSize) throw Exception.PayloadTooLarge(payloadLen, maxPayloadSize)

    val total = FrameHeader.LENGTH + payloadLen.toInt()
    if (bytes.size < total) return null

    val frame = bytes.copyOfRange(0, total)
    val rest = bytes.copyOfRange(total, bytes.size)
    buffer.reset()
    if (rest.isNotEmpty()) buffer.write(rest)
    return frame
  }

  public val bufferedByteCount: Int get() = buffer.size()

  public fun reset() { buffer.reset() }
}

File: packages/gateway/swift/Sources/BridgethingGateway/FrameAccumulator.swift
import Foundation

public struct FrameAccumulator {
  public static let defaultMaxPayloadSize: Int = 8 * 1024 * 1024

  public enum Error: Swift.Error, Equatable {
    case invalidMagic(UInt16)
    case unsupportedVersion(UInt8)
    case unsupportedCompression(UInt8)
    case unsupportedEncoding(UInt8)
    case payloadTooLarge(UInt64, max: Int)
  }

  private static let compactThreshold = 64 * 1024

  public let maxPayloadSize: Int
  private var buffer: Data = .init()
  private var head: Int = 0

  public init(maxPayloadSize: Int = FrameAccumulator.defaultMaxPayloadSize) {
    self.maxPayloadSize = maxPayloadSize
  }

  public mutating func append(_ chunk: Data) {
    compact()
    buffer.append(chunk)
  }

  private mutating func compact() {
    guard head > 0 else { return }
    if head == buffer.count {
      buffer.removeAll(keepingCapacity: true)
      head = 0
    } else if head >= Self.compactThreshold {
      buffer.removeSubrange(buffer.startIndex ..< buffer.index(buffer.startIndex, offsetBy: head))
      head = 0
    }
  }

  public mutating func nextFrame() throws -> Data? {
    guard buffer.count - head >= FrameHeader.length else { return nil }

    let at = head
    let header = buffer.withUnsafeBytes { raw -> Result<(UInt64, Compression, Encoding), Error> in
      let b = raw.bindMemory(to: UInt8.self)
      let magic = UInt16(b[at]) << 8 | UInt16(b[at + 1])
      guard magic == FrameHeader.magic else { return .failure(.invalidMagic(magic)) }
      guard b[at + 2] == FrameHeader.version else { return .failure(.unsupportedVersion(b[at + 2])) }
      guard let comp = Compression(rawValue: b[at + 3]) else {
        return .failure(.unsupportedCompression(b[at + 3]))
      }
      guard let enc = Encoding(rawValue: b[at + 4]) else {
        return .failure(.unsupportedEncoding(b[at + 4]))
      }
      var len: UInt64 = 0
      for i in (at + 8) ..< (at + 16) {
        len = (len << 8) | UInt64(b[i])
      }
      return .success((len, comp, enc))
    }

    let (payloadLen, _, _) = try header.get()
    if payloadLen > UInt64(maxPayloadSize) {
      throw Error.payloadTooLarge(payloadLen, max: maxPayloadSize)
    }

    let total = FrameHeader.length + Int(payloadLen)
    guard buffer.count - head >= total else { return nil }

    let start = buffer.index(buffer.startIndex, offsetBy: head)
    let frame = buffer.subdata(in: start ..< buffer.index(start, offsetBy: total))
    head += total
    compact()
    return frame
  }

  public var bufferedByteCount: Int { buffer.count - head }

  public mutating func reset() {
    buffer.removeAll(keepingCapacity: true)
    head = 0
  }
}
